---
title: "CellPlotManual"
author: "Sven E. Templer, Robert Sehlke"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{CellPlotManual}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Description

**CellPlot** is a [GNU R](http://r-project.org) package with plot methods for the
integrated visualisation of functional term enrichment and expression data. 
It can be used to display commonly used gene ontology (GO) term enrichment 
analysis from differential expression of genes (DEG) studies.

The development platform of this software is hosted on 
[github](https://github.com/dieterich-lab/CellPlot). 
In case of any questions, consider using the 
[issues](https://github.com/dieterich-lab/CellPlot/issues) system.

# Usage

## Load package and data

After installing the package (see [Installation](#installation) section), load it in a running
**R** session:

```{r}
library(CellPlot)
library(miscset) # only for examples in the vignette
```

The example data is available with the commands:

```{r}
data("golubGO")
data("leukemiasGO")
```

Reduce set of GO terms to display using the base **R** function `subset`:

```{r}
# select from the golubGO data:
x <- subset(golubGO$golub, pvalCutOff <= 0.05 & Significant > 5 & !duplicated(PROBEID))
x$Enrichment <- x$Significant / x$Expected
x <- sort(x, decreasing = TRUE, by = "Enrichment")
x <- head(x, 10)
# select from the leukemiasGO data:
y <- lapply(leukemiasGO, function (x) {
  x$Enrichment <- x$Significant / x$Expected
  x$Upregulated <- sapply(x$log2FoldChange, function (y) sum(y>0))
  x$Downregulated <- sapply(x$log2FoldChange, function (y) sum(y<0))
  x
})

```

## Cell Plot

The function `cell.plot` plots a horizontal barchart of strictly positive values in x. 
For each entry, a vector of additional values needs to be provided in a list. 
The additional values are plotted as cells subdividing the length of the corresponding bar. 
A typical usage scenario is plotting the enrichment of Gene Ontology terms, 
with individual cells reflecting the differential regulation of the constituent genes.

```{r, eval=FALSE}
cell.plot(x = setNames(x$Enrichment, x$Term), cells = x$log2fc,
          main ="GO enrichment on ALL/AML gene expression")
```
```{r, fig.width=6, fig.height=8, echo=FALSE}
cell.plot(x = setNames(x$Enrichment, x$Term), cells = x$log2fc, 
          cell.bounds = c(-6,6), main ="GO enrichment on ALL/AML gene expression", 
          x.mar = c(.4, 0), key.n = 7, y.mar = c(.1,0), cex = 1.6, cell.outer = 3, bar.scale = .7, space = .2)
```

## Sym Plot

The function `sym.plot` plots a split barchart, showing the proportions of 
two mutually exclusive sets in relation to a set containing them both. 
E.g., Gene Ontology terms, showing the proportions of differentially down-regulated 
and up-regulated annotated genes from a perturbation experiment. 
The color of the central column elements maps to the value provided in x (e.g. GO term enrichment). 
Associated genes may be provided as a list of vectors of expression values, 
same as for cell.plot(), or as separate vectors x.up and x.down, 
providing the numbers of up- and down-regulated genes in the same order as x.

```{r, eval=FALSE}
sym.plot(x = setNames(x$Enrichment, x$Term), cells = x$log2fc, x.annotated = x$Annotated, 
         main ="GO enrichment on ALL/AML gene expression")
```
```{r, fig.width=6, fig.height=6, echo=FALSE}
sym.plot(x = setNames(x$Enrichment, x$Term), cells = x$log2fc, x.annotated = x$Annotated, 
         main = "GO enrichment on ALL/AML gene expression",
         x.mar = c(.47, 0), key.n = 7, cex = 1.6, axis.cex = 0.8, group.cex = 0.7 ) 
```

## Arc Plot

This section will be updated/integrated in future.

```{r, fig.width=7, fig.height=6, eval=FALSE}
x$deg.up <- lapply(Map(setNames, x$deg.log2fc, x$genes), function (i) { i[i>0] })
x$deg.down <- lapply(Map(setNames, x$deg.log2fc, x$genes), function (i) { i[i<0] })
arc.plot( x = setNames(x$loge, x$term), up.list = x$deg.up, down.list = x$deg.down, x.mar=c(1,0.5))
```
```{r, fig.width=7, fig.height=6, echo=FALSE, eval=FALSE}
x$deg.up <- lapply(Map(setNames, x$deg.log2fc, x$genes), function (i) { i[i>0] })
x$deg.down <- lapply(Map(setNames, x$deg.log2fc, x$genes), function (i) { i[i<0] })
arc.plot( x = setNames(x$loge, x$term), up.list = x$deg.up, down.list = x$deg.down, x.mar=c(1,0.5))
```

## Go Histogram

This section will be updated/integrated in future.

```{r, fig.width=7, fig.height=6}
print(letters)
```

## Visual parameters

Both functions provide a range of parameters to tweak the visual appearance of the plots.

* **Font sizes** of the various elements can be adjusted using the `cex` parameters.
* **Internal margins** are adjusted through `x.mar` and `y.mar`. Used in particular to accomodate the label text to the left-hand side.
* **Absolute scaling** can be achieved through the `bar.scale` parameter, which acts as a multiplier for a predetermined average bar height. This is meant to be used in conjunction with a graphics device of constant size, to ensure visual consistency among multiple plots that vary in the number of terms displayed.
* **Data ranges** for all output may be fixed:
    * `elem.bounds` Require the cardinality of a term or significant subset to be in a specific range, e.g. `c(10,100)` -- between 10 and 100 genes.
    * `x.bound` Upper limit of the value displayed on the x-axis. For `sym.plot` this is must be a value between 0 and 100. The lower limit is always 0.
    * `cell.bounds` (only `cell.plot`) Sets the colour scale for cell data to a fixed range. If the plot contains values outside of that range, an indicator is added to the colour legend. This is particularly useful in the event of a few extreme outliers.
    * `mid.bounds` (only `sym.plot`) Similar to `cell.bounds`, this sets the colour range of the central column cells.
    * `sym` (only `cell.plot`), when set to `TRUE`, makes the scale of cell values symmetrical.
* **Colour schemes** may be controlled through character vectors provided to the following parameters. Internally, the R native function `ColorRampPalette()` is used.
    * `cell.col` (only `cell.plot`) Specify three valid colour names for cell data visualisation (lower extreme, mid point, upper extreme).
    * `mid.col` (only `sym.plot`) Specify two valid colour names to define the range of the central column cell colours. 
* **Colour keys** can be adjusted in their resolution (number of boxes) using the `key.n` parameter.
* **Gridlines** may be hidden via the `gridlines` parameter.

# Workflows

Gene ontology term enrichment in differential gene expression data can be 
performed using the 
[Bioconductor/topGO](http://www.bioconductor.org/packages/release/bioc/html/topGO.html)
package.

## Dataset `golubGO`

This section provides code and comments on the workflow to perform differential
expression of genes (DEG) analysis and the gene ontology term (GO) enrichment
of from the results applied to the `golub` dataset from the 
[Bioconductor/multtest](http://www.bioconductor.org/packages/release/bioc/html/multtest.html)
package.

It contains microarray gene expression data from leukemia study of *Golub* et
al. (1999), processed as described in *Dudoit* et al. (2002). Differential gene
expression was performed using a Student t-test to compare the two groups. GO 
annotation was done with the 
[Bioconductor/annotate](http://www.bioconductor.org/packages/release/bioc/html/annotate.html)
and
[Bioconductor/hu6800.db](http://www.bioconductor.org/packages/release/data/annotation/html/hu6800.db.html)
packages. 

```{r, eval=FALSE}
# CRAN
library(parallel)
library(plyr)
library(miscset)
# Bioconductor
library(topGO)
library(annotate)
library(hu6800.db)
library(multtest)

data(golub)

A <- split(golub, seq(nrow(golub)))
DEG <- do.call(rbind, mclapply(A, function(x){
  t <- t.test(x[as.logical(golub.cl)], x[!as.logical(golub.cl)])
  t$fc <- t$estimate[1]/t$estimate[2]
  suppressWarnings(t$fc <- log2(t$fc))
  data.frame(
    mean.aml = t$estimate[1],
    mean.all = t$estimate[2],
    log2fc  = t$fc,
    p = t$p.value,
    row.names = NULL)
}, mc.cores = 3))
DEG$padj <- p.adjust(DEG$p)
DEG <- data.frame(gene = golub.gnames[,3], DEG, stringsAsFactors = F)
DEG <- subset(DEG, !is.na(log2fc) & !is.na(padj))

M <- select(hu6800.db, DEG$gene, c("PROBEID","GO"))
genes <- dlply(subset(M, ONTOLOGY == "BP"), "PROBEID", function (x) unique(x$GO))
DEG <- subset(DEG, gene %in% names(genes))

## topGO object
GO <- new("topGOdata", ontology = "BP", description = 'golub',
          allGenes = setNames(DEG$padj, DEG$gene),
          geneSelectionFun = function (allScore) { allScore <= 0.05 },
          annotationFun = annFUN.gene2GO, gene2GO = genes)

GOsig <- lapply(list(golub=GO), function (x) {
  t <- new("elimCount", testStatistic = GOFisherTest, name = "Fisher test")
  s <- getSigGroups(x, t)
  GenTable(x, pvalCutOff = s, topNodes = length(x@graph@nodes))
})

golubGO <- Map(CellPlot::mergeGOdeg, GOsig, list(DEG), list(M), map.gene = "PROBEID", deg.p = "padj",deg.lfc="log2fc")
golubGO <- lapply(golubGO, subset, !is.na(PROBEID))
```

## Dataset `leukemiasGO`

This section provides code and comments on the workflow to perform differential
expression of genes analysis and the gene ontology term enrichment
of from the results applied to the `leukemiasEset` dataset from the 
[Bioconductor/leukemiasEset](http://www.bioconductor.org/packages/release/data/experiment/html/leukemiasEset.html)
package.

```{r, eval=FALSE}
# CRAN
library(parallel)
library(dplyr)
library(stringr)
# Bioconductor
library(BiocParallel)
library(DESeq2)
library(topGO)
library(annotate)
library(hu6800.db)
library(leukemiasEset)

data(leukemiasEset)

# M <- select(hu6800.db, featureNames(leukemiasEset), c("ENSEMBL","GO"), keytype = "ENSEMBL")
# M <- subset(M, ONTOLOGY == "BP", c("ENSEMBL","GO"))
# M <- M[!duplicated(M),]
# M <- dlply(M, "ENSEMBL", function (x) unique(x$GO))

# M <- as.character(unique(unlist(select(org.Hs.eg.db,featureNames(leukemiasEset),"ENSEMBL",keytype = "ENSEMBL"))))

A <- as(leukemiasEset,"data.frame")
A <- subset(A, LeukemiaType %in% c("NoL","ALL", "AML", "CLL"))
As <- subset(A, select = "LeukemiaType")
As$LeukemiaType <- relevel(factor(as.character(As$LeukemiaType)), "NoL")
A <- subset(A, select = grepl("^ENS", colnames(A)))
A <- sapply(A, as.integer)
A <- t(A)

DEG <- DESeqDataSetFromMatrix(countData = A, colData = As, design = ~ LeukemiaType)
DEG <- DESeq(DEG, fitType = "mean", parallel = T, BPPARAM = MulticoreParam(20))

n <- levels(As$LeukemiaType)[-1]
leukemiasGO <- lapply(setNames(n, n), function (n) {
  x <- results(DEG, c("LeukemiaType", n, "NoL"), "LeukemiaType", alpha = .05)
  x <- as.data.frame(x)
  x$padj[is.na(x$padj)] <- 1
  g <- new("topGOdata", ontology = "BP", description = 'Leukemia', 
           allGenes = setNames(x$padj, rownames(x)), 
           mapping = "org.Hs.eg.db",
           geneSelectionFun = function (allScore) { allScore <= 0.05 },
           annotationFun = annFUN.org, ID = "Ensembl")
  t <- new("elimCount", testStatistic = GOFisherTest, name = "Fisher test") # test definition
  s <- getSigGroups(g, t) # run F-test
  r <- GenTable(g, pvalCutOff = s, topNodes = length(g@graph@nodes)) # return data.frame
  r$pvalCutOff <- as.numeric(str_replace_all(r$pvalCutOff, "[^0-9e\\-\\.]*", ""))
  r$LogEnrich <- log2(r$Significant / r$Expected)
  ga <- genesInTerm(g) # GenesAnnotated | list of genes per go-terms
  ga <- ga[r$GO.ID] # eliminate missing terms
  names(ga) <- NULL
  r$GenesAnnotated <- ga
  xs <- x[,c("padj", "log2FoldChange")] # significant stats subset
  xs <- subset(xs, padj < 0.05)
  r$GenesSignificant <- lapply(r$GenesAnnotated, intersect, rownames(xs)) # extract genes
  ei.rows <- mclapply(r$GenesSignificant, function (y) {
    if (length(y)) as.list(xs[y,,drop=FALSE])
    else as.list(rep(NA_real_, length(xs)))
  }, mc.cores = 10)
  ei <- mclapply(names(xs), function(z) {
    lapply(ei.rows, "[[", z)
  }, mc.cores = 10)
  ei <- structure(ei, names = names(xs), row.names = seq(nrow(r)), class = "data.frame")
  row.names(ei) <- NULL
  r <- data.frame(r, ei, stringsAsFactors = FALSE, check.names = FALSE)
  return(r)
})

# lapply(GO, function(x) {list(all(r$Annotated == sapply(r$GenesAnnotated, length)),
#                              all(r$Significant == sapply(r$GenesSignificant, length)))})
leukemiasGO.ori <- leukemiasGO
leukemiasGO <- lapply(leukemiasGO, function(x) subset(x, pvalCutOff < 0.1))
save(leukemiasGO, file = "data/leukemiasGO.rdata")
```

# Installation

## CellPlot

Install the latest version from the repository on
[github.com/dieterich-lab/CellPlot](https://github.com/dieterich-lab/CellPlot). 
With the `devtools` package, it is an easy task:

```{r, eval=FALSE}
install.packages("devtools")
library(devtools)
install_github('dieterich-lab/CellPlot', build_vignettes = TRUE)
```

## Dependencies

```{r, eval=FALSE}
# load bioconductor functions
source("https://bioconductor.org/biocLite.R")

# install packages for CellPlot
install.packages("plyr")
biocLite("topGO")

# install packages for vignette, datasets, etc.
install.packages("miscset")
biocLite("leukemiasEset")
biocLite("multtest")
biocLite("hu6800.db")
```
